### DISCUSSION TITLE
How to turn on error msgs in the main loop?

### DISCUSSION THREAD

**Discussion by nussfluss** [dls-controls/pythonSoftIOC#93] [Q&A] [answered]
Hello,

I have recently started using pythonSoftIOC. Is there a way to display Python error msgs once ` softioc.interactive_ioc(globals())` has been called? `softioc.softioc.eltc(1)` does not seem to change anything.

cheers
nuss

**Comment by Araneidae** [comment]
I don't think I understand this question.  What error messages are you missing?

All `interactive_ioc` does is run a Python interactive shell, it is no more than a thin wrapper over `code.interact(...)`.  If you don't need this any other blocking operation will do, such as `cothread.WaitForQuit()` if you're using `cothread`, I expect there's an `aioca` equivalent.

**Reply by nussfluss** [reply]
Thx for the fast reply! I made a minimum code snippet showing my problem from your asyncio example (see **script#1**). I only added a single line. My issue is that the stack traceback is not printed to stdout when the Exception occurs (see **terminal#1**). That makes debugging soft IOCs really difficult. When I call the function `update()` from the interactive console, I get the correct error stack though (see **terminal#2**).
 
**script#1**:
```
# Import the basic framework components.
from softioc import softioc, builder, asyncio_dispatcher
import asyncio

# Create an asyncio dispatcher, the event loop is now running
dispatcher = asyncio_dispatcher.AsyncioDispatcher()

# Set the record prefix
builder.SetDeviceName("MY-DEVICE-PREFIX")

# Create some records
ai = builder.aIn('AI', initial_value=5)
ao = builder.aOut('AO', initial_value=12.45, always_update=True,
                  on_update=lambda v: ai.set(v))

# Boilerplate get the IOC started
builder.LoadDatabase()
softioc.iocInit(dispatcher)

# Start processes required to be run after iocInit
async def update():
    while True:
        ai.set(ai.get() + 1)
        raise Exception         <---------------------------------------------------------- line added
        await asyncio.sleep(1)

asyncio.run_coroutine_threadsafe(update(), dispatcher.loop)

# Finally leave the IOC running with an interactive shell.
softioc.interactive_ioc(globals())
```

**terminal#1**:
```
[stephan@eg8750 bpm-anomaly-detection]$ python reproduce_problem.py 
Starting iocInit
############################################################################
## EPICS 7.0.6.1
## Rev. 7.0.6.99.2.0
############################################################################
cas warning: Configured TCP port was unavailable.
cas warning: Using dynamically assigned TCP port 36139,
cas warning: but now two or more servers share the same UDP port.
cas warning: Depending on your IP kernel this server may not be
cas warning: reachable with UDP unicast (a host's IP in EPICS_CA_ADDR_LIST)
iocRun: All initialization complete
2022-06-21T10:16:14.067 Using dynamically assigned TCP port 42421.
Python 3.10.5 (main, Jun  6 2022, 18:49:26) [GCC 12.1.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
```

**terminal#2**:
```
>>> asyncio.run(update())
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/home/stephan/programs/python_packages/bpm-anomaly-detection/reproduce_problem.py", line 24, in update
    raise Exception
Exception
```

**Reply by coretl** [reply, marked as answer]
This is a failing of the example code. We put some exception handling in the dispatcher, so if an function passed to `on_update` fails we get the traceback printed, but we don't do anything with the long running `update` function in the example. We can make use of the same exception handling by scheduling it with the dispatcher:
```python
...
# Start processes required to be run after iocInit
async def update():
    while True:
        ai.set(ai.get() + 1)
        await asyncio.sleep(1)
        raise Exception("Boom")

dispatcher(update)

# Finally leave the IOC running with an interactive shell.
softioc.interactive_ioc(globals())
```
This now prints
```
Starting iocInit
############################################################################
## EPICS 7.0.6.1
## Rev. 7.0.6.99.2.0
############################################################################
iocRun: All initialization complete
Python 3.7.2 (default, Jan 20 2020, 11:03:41) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> ERROR:root:Exception when awaiting callback
Traceback (most recent call last):
  File "/dls/science/users/tmc43/common/python/pythonSoftIOC/softioc/asyncio_dispatcher.py", line 35, in async_wrapper
    await ret
  File "example_asyncio_ioc.py", line 26, in update
    raise Exception("Boom")
Exception: Boom
```
The logging message isn't the best though, I'll make an issue to track this

**Reply by nussfluss** [reply]
> dispatcher(update)

Thank you very much! That works indeed very well!!! :+1: I do not quite understand what the dispatcher call does but I am only getting started with asynchronous coding. So maybe I will get it in the future.

**Reply by nussfluss** [reply]
> future = asyncio.run_coroutine_threadsafe(update(), dispatcher.loop)
> print(future.result())

Thanks to you, as well!! That also works! :+1: The previous solution seems simpler though and I will stick with it if there is no catch.

**Reply by AlexanderWells-diamond** [reply]
> > future = asyncio.run_coroutine_threadsafe(update(), dispatcher.loop)
> 
> Thanks to you, as well!! That also works! +1 The previous solution seems simpler though and I will stick with it if there is no catch.

There's no catch to it, I realised that my suggestion was actually worse than coretl's and so I deleted it.

**Reply by nussfluss** [reply]
Well it worked and I am glad that you replied!

**Reply by coretl** [reply]
> I do not quite understand what the dispatcher call does but I am only getting started with asynchronous coding

Calling `dispatcher(func)` calls this code: https://github.com/dls-controls/pythonSoftIOC/blob/master/softioc/asyncio_dispatcher.py#L30-L38

The `run_coroutine_threadsafe` call is the same as what is in the example, the difference is that `func` is called within a wrapper that catches errors and reports them

**Reply by nussfluss** [reply]
Okay, thx for the info! That clarfies it indeed a little.
